dist/tools/ci/can_fast_ci_run.py

```diff
@@ -0,0 +1,223 @@
+#!/usr/bin/python3
+"""
+Command line utility to check if only a subset of board / application combinations
+need to be build in the CI
+"""
+import argparse
+import io
+import json
+import os
+import re
+import subprocess
+import sys
+
+REGEX_GIT_DIFF_RENAME_COMPLEX = re.compile(r"^[0-9-]+\s+[0-9-]+\s+(.*)\{(.*) => (.*)\}(.*)$")
+REGEX_GIT_DIFF_RENAME_SIMPLE = re.compile(r"^[0-9-]+\s+[0-9-]+\s+(.*) => (.*)$")
+REGEX_GIT_DIFF_SINGLE_FILE = re.compile(r"^[0-9-]+\s+[0-9-]+\s+(.*)$")
+
+# Beware: Order matters. The first matching rule will be applied
+OTHER_CLASSIFIERS = [
+    [re.compile(r"^(Makefile.*|sys\/Makefile.*|drivers\/Makefile.*|makefiles\/.*)$"), "build-system"],
+    [re.compile(r"^(drivers\/include\/.*|sys\/include\/.*)$"), "public-headers"],
+    [re.compile(r"^(Kconfig|kconfigs\/.*|pkg\/Kconfig|sys\/Kconfig|drivers\/Kconfig)$"), "kconfig"],
+    [re.compile(r"^(.*\.cff|doc\/.*|.*\.md|.*\.txt)$"), "doc"],
+    [re.compile(r"^(CODEOWNERS|.mailmap|.gitignore|.github\/.*)$"), "git"],
+    [re.compile(r"^(\.murdock|dist\/ls\/.*|\.drone.yml)$"), "ci-murdock"],
+    [re.compile(r"^(\.bandit|\.circleci\/.*|\.drone.yml)$"), "ci-other"],
+    [re.compile(r"^(dist\/.*|Vagrantfile)$"), "tools"],
+]
+
+REGEX_MODULE = re.compile(r"^(boards\/common|core|cpu|drivers|sys)\/")
+REGEX_PKG = re.compile(r"^pkg\/")
+REGEX_BOARD = re.compile(r"^boards\/")
+REGEX_APP = re.compile(r"^(bootloaders|examples|fuzzing|tests)\/")
+
+EXCEPTION_MODULES = {"boards/common/nrf52"}
+
+
+def print_change_set_section(name, contents):
+    """
+    Print the given change set section human reable
+    """
+    if len(contents) > 0:
+        print(name, file=sys.stderr)
+        print("=" * len(name), file=sys.stderr)
+        print("", file=sys.stderr)
+        for category in sorted(contents):
+            print(category, file=sys.stderr)
+            print("-" * len(category), file=sys.stderr)
+            print("", file=sys.stderr)
+            for file in sorted(contents[category]):
+                print(f"- {file}", file=sys.stderr)
+            print("", file=sys.stderr)
+
+
+class ChangeSet:
+    """
+    Representation of the modules affected by a change set
+    """
+    def __init__(self, riotbase=None):
+        if riotbase is None:
+            riotbase = os.getcwd()
+        self.apps = {}
+        self.boards = {}
+        self.modules = {}
+        self.other = {}
+        self.pkgs = {}
+        self.__riotbase = os.path.normpath(riotbase)
```

Comment:
In Pep8, the convention for internal use attribute is to use a single leading underscore:

```
_single_leading_underscore: weak "internal use" indicator. E.g. from M import * does not import objects whose names start with an underscore.
```